### ISSUE TITLE
.NET 6 | Support the new BCL DateOnly and TimeOnly structs

### ISSUE THREAD

**Issue by roji** [dotnet/SqlClient#1009]
New DateOnly and TimeOnly structs are being introduced to .NET 6 as alternatives to DateTime (https://github.com/dotnet/runtime/issues/49036). SqlClient and add support for these as perfect mappings for SQL Server `date` and `time`.

**Comment by Wraith2** [comment]
Are they likely to have ADO surface area, e.g. `DbDataReader.GetDateOnly` etc or do you think that providing only `GetFieldValue(Async)<T>` is appropriate because that wouldn't need any new methods?

**Comment by mattjohnsonpint** [comment]
Given that the SQL type is `date`, I suggest making a method `DBDataReader.GetDate` even if the type it returns in .NET is called `DateOnly` (which is still being discussed).

**Comment by Wraith2** [comment]
@mattjohnsonpint I think that would depend on the ranges supported. If it maps directly then I agree if there are any differences we may want to call out the DataOnly type is not the same as sql DATE. Mapping for clr to sql types is already complicated and a bit messy imo I think we should take care not to make it worse for the save of brevity.

**Comment by mattjohnsonpint** [comment]
They have the exact same range and precision.

**Comment by cheenamalhotra** [comment]
Acknowledged.
Looking forward to solidification in design.

**Comment by roji** [comment]
> Are they likely to have ADO surface area, e.g. DbDataReader.GetDateOnly etc or do you think that providing only GetFieldValue(Async)<T> is appropriate because that wouldn't need any new methods?

I'm not aware of a reason to add new GetDate/GetTime APIs to DbDataReader - the generic GetFieldValue (and Async) should be sufficient for retrieving the new types (unless you guys have a reason?).

**Comment by mattjohnsonpint** [comment]
No strong opinions here, but a reason would be discoverability for those used to APIs like `DbDataReader.GetDateTime`.

**Comment by roji** [comment]
> No strong opinions here, but a reason would be discoverability for those used to APIs like DbDataReader.GetDateTime.

That's true, but don't think it's sufficient... Note how there are no async equivalents for the Get* methods (e.g. no GetDateTimeAsync), and users must use the generic GetFieldValueAsync. Basically I think the Get* methods are a bit of a remnant from the days before generics existed.

**Comment by mattjohnsonpint** [comment]
Please note that the final names for these types are `DateOnly` and `TimeOnly` and that they are now merged into the main branch for the next version of .NET 6 (preview 4 likely).  Please rename the title of this issue accordingly.  Thanks.

**Comment by roji** [comment]
FYI preview4 is out and these types are now available (I've added support in Npgsql). If you guys can add support at your level, I can do that in EF Core for SQL Server as well.

**Comment by CleanCodeX** [comment]
"If you guys can add support at your level, I can do that in EF Core for SQL Server as well."

What's the current status here? It seems, DateOnly support ist not given yet (in NET 6 preview 6)

**Comment by Wraith2** [comment]
Did you mean to ask in EFCore?

Status in SqlClient is that it's waiting for the NET6 release, CI update, new project targets, then work, then testing, then release which is every 6 months.

**Comment by ErikEJ** [comment]
TBH, do not expect anything to be available until EF Core 7 (November 2022)

**Comment by CleanCodeX** [comment]
> Did you mean to ask in EFCore?
Yes EF Core.

If it takes that long to support it, then how can it be used then in Models without canging the property type? Can a conversion (HaveConversion/HasConversion) be used? 

This:

`builder.Properties<DateOnly>().HaveConversion(typeof(DateTime));`

does not work.

**Comment by John0King** [comment]
@ErikEJ  
> TBH, do not expect anything to be available until EF Core 7 (November 2022)

what make SqlClient's development this kind of slow ?
and why can not just release it now ?   from above discussion , I can not see any  issue block the development

**Comment by Wraith2** [comment]
To start with .net 6 hasn't been released yet. And then there's all the work involved that I mentioned above., CI update, new project targets, implement, testing, documentation, then release. If you want to help with that then go ahead.

**Comment by NetMage** [comment]
> To start with .net 6 hasn't been released yet. And then there's all the work involved that I mentioned above., CI update, new project targets, implement, testing, documentation, then release. If you want to help with that then go ahead.

Why does an update to SqlClient need to start after the FTM of .Net 6? Why didn't the development start in conjunction with .Net 6 preview and then be ready to release the next available release after .Net 6 RTM (which must be before Nov 22)? Is the release schedule listed anywhere?

**Comment by Wraith2** [comment]
Because the MS team is small and adding this feature wasn't a priority. It can't be added by entirely be an external contributor because all the work requires review by the MS team prior to merging. There are also very few external contributors because most people seem to prefer complaining to helping.

**Comment by musictopia2** [comment]
Any idea of when DateOnly and TimeOnly will be supported in SQL Server.  This can cause many problems in businesses if they need the new DateOnly and TimeOnly using SQL Server.

**Comment by CleanCodeX** [comment]
"Because the MS team is small and adding this feature wasn't a priority."
A small "MS"-Team? Why is the whole (I guess) Microsoft-Team small? Because of the pandemic or was it already small before? I mean, it's a multi billion company which spent more than a billion for developing .net. It surprises that the team is too small to integrate Date/TimeOnly support for MS's own product SQL-Server.

**Comment by Wraith2** [comment]
It's an inaccurate perspective that people have of microsoft as a single immense monolithic entity. As you follow the open source projects here on github you realise that there are actually fairly small teams of people working on these projects. Even runtime which is about as big as it gets isn't nearly as many people as you'd expect. 

Why is the team small? Because that's what is seen as needed, I expect. As an outsider I speculate that since there is no direct revenue stream from this library (or other drivers) there is no particular incentive to hire more people to deal with it. Having a team of support people for the driver is probably a loss offset again the sql server licensing so why wouldn't the business want as few people as possible?

Now the next question you're likely to ask is why isn't this handled with the same sort of gung-ho spirit as the runtime and aspnet products which are also open source? I think answer to that is it's a different division inside MS with different priorities. How do you make this feature a priority? If you're not paying then you're probably stuck with the usual visual studio feedback channels or social media rabble rousing. If you're a business then try reaching out through whatever channels are available and use your wallet clout to back up your feedback.

**Comment by ErikEJ** [comment]
Also, .NET 6 was only released few days ago, and I already see indications of efforts to support .NET 6.

**Comment by Joren-Thijs-KasparSolutions** [comment]
@ErikEJ thats good but it would be great if in the future this kind of work is done beforehand.

MS advertised the handy new `DateOnly` and `TimeOnly` fields as part of .NET 6 so you don't need to store any unneeded info.
But the problem everyone is running into is that they can't store it anywhere because MS their own storage technologie like SQL server client or EF Core can't handle it.

**Comment by CleanCodeX** [comment]
> @ErikEJ thats good but it would be great if in the future this kind of work is done beforehand.
> 
> MS advertised the handy new `DateOnly` and `TimeOnly` fields as part of .NET 6 so you don't need to store any unneeded info. But the problem everyone is running into is that they can't store it anywhere because MS their own storage technologie like SQL server client or EF Core can't handle it.

That's not really a problem as MS SQL Server *does* support Date and Time colum types. The only need here is to use a custom converter like this. 
There's also a nuget package from someone else for that.

All you need is:

```csharp
protected override void ConfigureConventions(ModelConfigurationBuilder builder)
        {
            builder.Properties<DateOnly>()
                .HaveConversion<DateOnlyConverter>()
                .HaveColumnType("date");

            builder.Properties<DateOnly?>()
                .HaveConversion<NullableDateOnlyConverter>()
                .HaveColumnType("date");
        }
```

It would be nice if this comparer could be integrated into EF Core, so people would only declare something like this:

```csharp

    protected override void ConfigureConventions(ModelConfigurationBuilder builder)
    {
        builder.Properties<DateOnly>().HaveConversion<DateTime>();
    }
```

**Comment by roji** [comment]
@CleanCodeX that's an EF Core solution - SqlClient needs to support DateOnly and TimeOnly in order for non-EF users to be able to work with it.

**Comment by Joren-Thijs** [comment]
@CleanCodeX i did see your converter solution in the ef core issue and i do appreciate it but is still sad it's not supported out of the box.

The problem is not that i have a problem implementing a simple converter, the problem is my collegeau ran into this when trying to update to .net 6. As a consequence he now doubts the quality of this framework leading to more pointless discussions around upgrading and witch tech stack to use...

**Comment by musictopia2** [comment]
The problem is this only works if you are using the ef core.  Does not work if you are using dapper.

**Comment by AbstractionsAs** [comment]
I'm currently all kinds of blicked by this. I'm not using EF, so EF solutions are no help..

**Comment by Wraith2** [comment]
How are you blocked? These are new types that can be replaced with a little work using the existing DateTime type. No timescale has been given for supporting these types so there is no missed deadline that you were expecting anyone to hit.

**Comment by aQsu** [comment]
I guess coders have been waiting for sane date and time related types for years . When they are finally available, we want to use them.

**Comment by JRahnama** [comment]
@aQsu it is in our to do list. Hopefully if everything works as planned we will try to have this for version 5 release.

**Comment by Grauenwolf** [comment]
Will this be added to System.Data.SqlClient as well or is it only going to be a Microsoft.Data.SqlClient feature?

**Comment by ErikEJ** [comment]
@Grauenwolf System.Data.SqlClient will not receive any new features, only security updates.

**Comment by Grauenwolf** [comment]
> I'm currently all kinds of blocked by this. I'm not using EF, so EF solutions are no help..

I ended up building an adapter layer into my ORM, Tortuga Chain. I'm assuming the other ORM developers are going to have to do the same.

**Comment by mabster** [comment]
> @aQsu it is in our to do list. Hopefully if everything works as planned we will try to have this for version 5 release.

Hi @aQsu! I see that v5 was released a few days ago. Did this make the cut? I didn't see a mention of TimeOnly or DateOnly in the release notes.

**Comment by SimonCropp** [comment]
@mabster no use of `DateOnly` in the code https://github.com/dotnet/SqlClient/search?q=DateOnly&type=code
so i would assume no

**Comment by Pilchard123** [comment]
Of course, that now raises the question "when is it likely to be done"? Is there an agreed API that is just waiting on someone to make a PR, or is it still in the planning stage? Is it something you'd accept a PR for if there was one?

**Comment by Wraith2** [comment]
The discussion above indicates that the expected additional surface area is to GetFieldValue<T> and GetFieldValueAsync<T>. Other places like parameters and the internal SqlBuffer types will need updating to handle the extra data type but those aren't externally visible.

There is currently no build of this library that targets anything beyond netcore 3.1. As soon as there is one then the new functionality can be added. Until then it will wait. There is no indication of where this is in terms of priority but there are other tasks like the merging of netfx and netcore codebases which imo should take priority over adding new features.

**Comment by nabenzine** [comment]
Any update on this ? is this included in Ef core 7 previews ?

**Comment by Wraith2** [comment]
No. There is no update. There is still no net6 build target for this library and that is still required to begin work on the feature. If there were an update I'm sure someone would post about it here.

**Comment by Grauenwolf** [comment]
You don't really need support for it at this level. You could handle it all in the ORM. (That's what my ORM did.)

I'm not even sure what support would look like. Suddenly changing columns that were materialized as DateTime to be DateOnly or TimeOnly would be a breaking change.

**Comment by MisinformedDNA** [comment]
You may not _need_ it, but it's handier then everyone adding their own value converters. I don't see why there would need to be a breaking change. If a user changes their code to use one of the new objects, then they are changing their own functionality.

**Comment by roji** [comment]
> I'm not even sure what support would look like. Suddenly changing columns that were materialized as DateTime to be DateOnly or TimeOnly would be a breaking change.

I think we discussed this above; support would mean allowing asking `GetFieldValue<T>` for a DateOnly/TimeOnly, and allowing SqlParameter.Value to be DateOnly/TimeOnly. This is completely non-breaking, and would also allow seamless support in ORMs layered over SqlClient.

These types are now a standard part of the .NET API, database drivers really should just support them at this point.

**Comment by Grauenwolf** [comment]
When you call GetValues, can DateOnly appear in the array for Date columns?

If yes, it's a breaking change and people will be mad.

If no, it doesn't faithfully represent what's in the database and people will be mad. 

Does it make sense to have a switch that controls this? Maybe, but it would be non-obvious and that's close to not existing. 

Hence my concern about not knowing what this is going to look like.

**Comment by MisinformedDNA** [comment]
> When you call GetValues, can DateOnly appear in the array for Date columns?

Based on previous comments, I think they are only going to support `GetFieldValue(Async)<T>`:

> > No strong opinions here, but a reason would be discoverability for those used to APIs like DbDataReader.GetDateTime.
> 
> That's true, but don't think it's sufficient... Note how there are no async equivalents for the Get* methods (e.g. no GetDateTimeAsync), and users must use the generic GetFieldValueAsync. Basically I think the Get* methods are a bit of a remnant from the days before generics existed.

**Comment by roji** [comment]
> When you call GetValues, can DateOnly appear in the array for Date columns?
> If yes, it's a breaking change and people will be mad.
> If no, it doesn't faithfully represent what's in the database and people will be mad.

@Grauenwolf, the Npgsql, Microsoft.Data.Sqlite, and MySqlConnector drivers have implemented this change. At least for the 1st two, this allows you to send DateOnly values in DbParameter, and use GetFieldValue<T> to read them. This isn't a breaking change, and absolutely nobody got mad that GetValues hasn't changed. Modern usage of ADO.NET generally uses GetFieldValue<T> anyway - it's generic, and so doesn't require additional casting and also doesn't box value type. In my experience supporting Npgsql users, not many still use the non-generic read APIs (GetValue/GetValues) - these really make sense mainly in fully dynamic scenarios, where your program sends arbitrary SQL queries and doesn't know what types will be coming back.

Sure, it's possible to add an opt-in to also change the "default" CLR type for dates from DateTime to DateOnly; that would indeed bring this change to the non-generic APIs as well. I've personally not seen one request for it so far, so I'd really start off with the non-breaking change (parameters and generic GetFieldValue), and see from there.

**Comment by AbstractionsAs** [comment]
> 

Exactly this

**Comment by FransBouma** [comment]
@roji 
> the Npgsql, Microsoft.Data.Sqlite, and MySqlConnector drivers have all implemented this change. At least for the first two, this allows you to send DateOnly values in DbParameter, and use GetFieldValue to read them. This isn't a breaking change, and absolutely nobody got mad that GetValues hasn't changed.

... still you changed it in npgsql 10 which broke my code. (as my code assumed the npgsql 9 types were still in play). 

I must say the inconsistency between reader.GetValue() and reader.GetFieldValue<T>() is pretty bad here: as I can save DateOnly values but fetching them with a reader and then use GetValues() gives me DateTime typed values, while using reader.GetFieldValue<T>() will give me DateOnly values; this isn't the case by any other type; all other types are consistent across these methods. 

I don't expect this to be changed, just wanted to mention that, hey, I'm one of the people who ran into this and can throw away a day's work. :) 

The main issue is that if fixed types are used in an entity class, e.g. a DateOnly and TimeOnly property in a class, mapped to Date and Time fields in the database through some layer, the mapping system has to make sure a value in the object can be written to the table and a value read from the table can be placed in the object. We use reader.GetValues() to get the values asap from the reader, and close the reader quickly. A typed projection generated from a lambda directly on the reader will use the reader.GetFieldValue<T>(), this is for linq queries etc. but not for entity fetches. 

The inconsistency is problematic; I can live with GetValues() returning a DateOnly type always, I have to make sure the mapped entity property is of type DateOnly as well. However if one projection pipeline using GetFieldValue<T> will produce DateOnly and another one will produce DateTime values, then a problem arises: the only solution then is to go for the type that has always worked: DateTime, but this limits the expression freedom for the entity type: the property can't be of type DateOnly.

**Comment by roji** [comment]
Hey @FransBouma 👋 

Can you help me understand exactly where the inconsistency is which you're seeing? Both `NpgsqlDataReader.GetValue()` and `GetValues()` are non-generic methods which return/populate `object` - these have both been changed to return .NET DateOnly for a PostgreSQL `date`. `GetFieldValue<T>()` returns whatever you ask of it: you can ask it for a DateOnly or a DateTime, Npgsql supports either one. See the quick code sample below that shows all this working.

If there's some sort of inconsistency/bug we can definitely fix it for 10.0.1.

BTW maybe this is better discussed on the Npgsql repo, do you want to open an issue there?

<details>
<summary>Code sample</summary>

```c#
var builder = new NpgsqlDataSourceBuilder("Host=localhost;Username=test;Password=test");
await using var dataSource = builder.Build();
await using var connection = await dataSource.OpenConnectionAsync();

await using var command = new NpgsqlCommand("SELECT '2020-03-01'::date", connection);
await using var reader = await command.ExecuteReaderAsync();
await reader.ReadAsync();

Console.WriteLine(reader.GetValue(0).GetType().Name); // DateOnly

var values = new object[1];
reader.GetValues(values);
Console.WriteLine(values[0].GetType().Name); // DateOnly

Console.WriteLine(reader.GetFieldValue<DateOnly>(0).GetType().Name); // DateOnly
Console.WriteLine(reader.GetFieldValue<DateTime>(0).GetType().Name); // DateTime
```

</details>

**Comment by FransBouma** [comment]
hey @roji !

Sorry for the confusion, I hope to clear it up: I referred to npgsql because in your post https://github.com/dotnet/SqlClient/issues/1009#issuecomment-1257150173 you mention that it's not really needed to change GetValue/GetValues behavior regarding DateOnly/TimeOnly while (albeit sometime later) you did change it for npgsql. 

The inconsistency is in Microsoft.Data.SqlClient, not npgsql, where it *is* consistent. The change you made in npgsql 10 does illustrate the problems that arise however: a breaking change in the ado.net provider means a breaking change in the middleware using the ado.net provider. If these are version-coupled (like entity framework) you won't notice. However many database middleware isn't: there the ado.net provider is loosely coupled e.g. through the DbProviderFactory. (I'm glad you added a switch to get the old behavior back in 10.0.1 btw!)

ADO.NET providers map database types to .net types and it's key to have a consistent mapping for *all* supported types, so a user knows what to expect when calling a function on a datareader: they *know* for database type A with precision/scale P and S they'll get a value V with type T back when calling a method on the datareader, doesn't matter which one!

Here, in Microsoft.Data.SqlClient that's *not* the case for DateOnly/TimeOnly. This is *BAD*. Changing the behavior for DateOnly/TimeOnly is a breaking change but IMHO this should be handled as follows:

- keep the old behavior as the default. (So map to DateTime and TimeSpan). The main reason is that users of a database middleware aren't aware of the breaking change nor should they be; they're using a middleware!
- add the new mapping (to DateOnly and TimeOnly) *consistently* across all methods and add a switch to enable this behavior, which basically means, switching to another mapping table for the .net <-> db type mapping

I've seen some breaking changes in ado.net providers in the past 23 years I'm now writing ORMs (Firebird, ODP.NET, Npgsql) and it's always a nightmare. The middleware developer might be able to cope with it, the users are caught off guard, as the breaking change ripples through. 

Thanks for considering

**Comment by roji** [comment]
> I referred to npgsql because in your post https://github.com/dotnet/SqlClient/issues/1009#issuecomment-1257150173 you mention that it's not really needed to change GetValue/GetValues behavior regarding DateOnly/TimeOnly while (albeit sometime later) you did change it for npgsql.
> 
> The inconsistency is in Microsoft.Data.SqlClient, not npgsql, where it is consistent.

Ah, I understand. Yeah, it seemed that enough time has passed for DateOnly/TimeOnly to propagate through the ecosystem, and also crucially, the new version of Npgsql no longer supports old TFMs where DateOnly/TimeOnly didn't yet exist (otherwise we'd have the same version of Npgsql returning different things based on TFM, which seems bad). FWIW we've had various debates between the maintainers if this breaking change was worth it, not everyone necessarily agrees; but I admit I'm more on the side of moving forward with the ecosystem, and DateOnly/TimeOnly are now the canonical/standard ways to represent date and time, so the change made sense to me.

But regardless, I'm still not exactly sure where SqlClient isn't consistent - can you provide a quick code sample illustrating the problem? AFAICT SqlClient always returns DateTime by default, except if you call `SqlDataReader.GetFieldValue<DateOnly>()`, in which case it returns DateOnly (see code sample just below which plays around with this). But I'm sure I'm missing something - can you point out the inconsistency?

<details>
<summary>SqlClient code sample</summary>

```c#
await using var connection = new SqlConnection("Server=localhost;Database=test;User=SA;Password=Abcd5678;Connect Timeout=60;ConnectRetryCount=0;Encrypt=false");
await connection.OpenAsync();

// Read side
await using (var command = new SqlCommand("SELECT CAST('2020-03-01' AS date)", connection))
{
    await using var reader = await command.ExecuteReaderAsync();
    await reader.ReadAsync();

    Console.WriteLine(reader.GetFieldType(0).Name); // DateTime
    Console.WriteLine(reader.GetValue(0).GetType().Name); // DateTime

    var values = new object[1];
    reader.GetValues(values);
    Console.WriteLine(values[0].GetType().Name); // DateTime

    Console.WriteLine(reader.GetFieldValue<DateOnly>(0).GetType().Name); // DateOnly
    Console.WriteLine(reader.GetFieldValue<DateTime>(0).GetType().Name); // DateTime
}

// Write side
await using (var command = new SqlCommand("SELECT @dateonly, @datetime", connection))
{
    command.Parameters.AddWithValue("@dateonly", new DateOnly(2020, 3, 1));
    command.Parameters.AddWithValue("@datetime", new DateTime(2020, 3, 1));

    await using var reader = await command.ExecuteReaderAsync();
    await reader.ReadAsync();

    Console.WriteLine(reader.GetFieldType(0).Name); // DateOnly
    Console.WriteLine(reader.GetFieldType(1).Name); // DateTime
}
```

</details>

> I'm glad you added a switch to get the old behavior back in 10.0.1 btw!

FWIW we'll probably end up not introducing a new switch, but rather just publish a fragment that users can paste in to alter Npgsql's mapping (we have APIs for that), to revert to DateTime/TimeSpan by default. That's just a minor detail (the important thing is that there'll be a way to revert back - even on 10.0.0 BTW).

> ADO.NET providers map database types to .net types and it's key to have a consistent mapping for all supported types, so a user knows what to expect when calling a function on a datareader: they know for database type A with precision/scale P and S they'll get a value V with type T back when calling a method on the datareader, doesn't matter which one!

That would certainly be ideal, but the unfortunate reality is that database types are, well, a total mess... SQL Server and PostgreSQL have completely different types around timestamps: SQL Server has datetime, datetime2 and datetimeoffset; PG has timestamp (timestamp with unknown/implicit timezone) and timestamptz (timestamp assumed to be UTC), with nothing that can hold a timezone. And don't even mention textual types (varchar/nvarchar with max or otherwise on SQL Server, compared to text on PG). IIRC Oracle doesn't have time (only interval), and of course SQLite has nothing at all...

So if an application or layer is handling both database types and the mapping between them and .NET types, I don't think they have a choice but to know about each database and how .NET types map to database types for that database. Though of course, adding a driver version to the mix (Npgsql 10 vs. older versions) definitely adds more complexity, I do get that 100%.

**Comment by FransBouma** [comment]
> But regardless, I'm still not exactly sure where SqlClient isn't consistent - can you provide a quick code sample illustrating the problem? AFAICT SqlClient always returns DateTime by default, except if you call SqlDataReader.GetFieldValue<DateOnly>(), in which case it returns DateOnly (see code sample just below which plays around with this). But I'm sure I'm missing something - can you point out the inconsistency?

As there's no implicit conversion from DateTime to DateOnly and vice versa, there's explicit code added to make `GetFieldValue<DateOnly>()` possible if the underlying type is actually DateTime (as reported by GetFieldType(ordinal)). Yes I know the code internally doesn't convert from DateTime to DateOnly but converts the raw data directly to DateOnly, but from the outside it looks like that. The core issue is that the answer to "to what type does SQL Server's type 'Date' map to in the .net space using Microsoft.Data.SqlClient?" is "it depends". While that's classic database-esk, it's also not what you want: you want a clear cut answer: a single type you can rely on the ado.net provider will return when a value comes back from the database: that value will be of that particular type and all code can depend on that. Here that's not clear cut; Microsoft.Data.SqlClient has 'some' support for DateOnly but not all the way; it didn't make the choice, only ... partly, which is basically `return DateOnly.FromDateTime(GetValue(ordinal))`. 

> db types are a mess

True, but within the scope of a single ADO.NET provider they should be consistent. Of course any ADO.NET provider makes their own choice, hence we have e.g. the lovely type disaster in ODP.NET for int32/int64 which even differ if you use it with EF Core or native!

Coming back to the point: as the Microsoft.Data.SqlClient team made *some* effort to have some support for DateOnly (GetFieldValue, datatable) but not make the tough choice to simply define the mapping properly, it's a confusing mess. But like I said, I don't expect it to change. At least not anytime soon.

**Comment by roji** [comment]
Ah, I see - the inconsistency you're pointing out is between what's returned by SqlDataReader.GetFieldType() (DateTime, also the default as returned by non-generic GetValue()) and using the generic `GetFieldValue<DateOnly>()` to fetch a DateOnly, as requested/wanted by the user.

> The core issue is that the answer to "to what type does SQL Server's type 'Date' map to in the .net space using Microsoft.Data.SqlClient?" is "it depends".

I'd phrase it slightly different - Sqlclient maps SQL Server `date` to DateTime by default, but allows also mapping it to DateOnly (non-default) if the user desires that (but I see what you mean). FWIW Npgsql (and I think most other ADO.NET providers) also has various cases where there's one default type, but the you can use `GetFieldValue<T>()` to get another non-default type ([see these docs](https://www.npgsql.org/doc/types/basic.html#read-mappings), where there's a column for the default and then another for other options).

I'll even point out that before Npgsql 10, Npgsql had the exact same behavior: the default for `date` was DateTime (so that's what GetFieldType() and non-generic GetValue() returned), but you could still request a DateOnly via generic `GetFieldValue<T>()`. And with Npgsql 10 the situation is the same, only reversed: you get DateOnly by default, but can still ask for a DateTime (so it's "inconsistent" in that sense).

So I'm basically pointing out that this situation isn't unique to SqlClient or date/DateOnly, and I'm not sure it's right to call this situation "inconsistent": it's just that there's always a single default value, and then optionally additional non-default ones.
